biovectors_modules/word2vec_analysis_helper.py

```diff
@@ -34,33 +34,70 @@ def generate_timeline(
     return timeline_df
 
 
-def get_global_distance(
-    year_one_model: str, year_two_model: str, shared_tokens: Sequence[str]
-) -> pd.DataFrame:
-    """
-    This function is designed to get the global distance of ever token between two years.
-    Global distance is basically the cosine similarity (token_year_one, token_year_two)
+def _calculate_distances(
+    year_one_model: pd.DataFrame,
+    year_two_model: pd.DataFrame,
+    neighbors: int,
+    input_queue: JoinableQueue,
+    output_queue: JoinableQueue,
+):
 
-    Parameters:
-        year_one_model - The first year to be compared
-        year_two_model - the second year to be compared
-        shared_tokens - the list of all tokens shared across the years
-    """
+    while True:
+        tok = input_queue.get()
 
-    distance_matrix = cdist(year_one_model, year_two_model, "cosine")
-    token_dist_iterator = zip(shared_tokens, np.diag(distance_matrix))
-    global_distance = pd.DataFrame.from_records(
-        [{"token": token, "global_dist": dist} for token, dist, in token_dist_iterator]
-    )
+        if tok is None:
+            break
+
+        w_t1 = (
+            year_one_model.query(f"token == {repr(tok)}").drop("token", axis=1).values
+        )
+        w_t2 = (
+            year_two_model.query(f"token == {repr(tok)}").drop("token", axis=1).values
+        )
 
-    return global_distance
+        model_one_neighbors = list(
+            map(lambda x: x[0], get_neighbors(year_one_model, tok, neighbors))
+        )
+        model_two_neighbors = list(
+            map(lambda x: x[0], get_neighbors(year_two_model, tok, neighbors))
+        )
 
+        # w_(t) and w_(t+1)
+        word_vec = np.vstack([w_t1, w_t2])
 
-def get_local_distance(
-    year_one_model: str,
-    year_two_model: str,
+        # N(w_(t)) union N(w_(t+1))
+        neighbor_vec = np.vstack(
+            [
+                year_one_model.query(f"token in {model_one_neighbors}")
+                .set_index("token")
+                .values,
+                year_two_model.query(f"token in {model_two_neighbors}")
+                .set_index("token")
+                .values,
+            ]
+        )
+
+        # Grab the similarity of word vectors
+        output_queue.put(
+            {
+                "token": tok,
+                "local_dist": pdist(
+                    1 - cdist(word_vec, neighbor_vec, "cosine"), "cosine"
+                ).item(),
```

Comment:
This is a bit hard to follow. If I understand correctly, you calculate the similarities between those two sets (`word_vec` and `neighbor_vec`), that is a word embedding representation (I think) and its neighbors, leading to an array (n x m). And then you calculate the distance (1 - sim) between those similarities (leading to n x n)? And this is the local distance, interesting! If that's correct, it was not hard to follow! :joy:

Reply:
Yup! You got it. The local distance is a bit wonky of a description but basically its the distance of a tokens; similarity to all neighbors within both time periods.